smolyay/samples.py

```diff
@@ -508,3 +508,620 @@ def _create(self):
+            self._create()
+            self._valid_cache = True
+        return self._indexes
+
+
+class TensorProductPointSet(PointSetProduct):
+    """Generate a grid using all combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        self._points = numpy.array(list(itertools.product(*self._point_sets)))
+
+
+class SmolyakSparseProductPointSet(PointSetProduct):
+    """Generate a grid using sparse combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        grid_points = None
+        max_num_level = max([ob.num_levels for ob in self._point_sets])
+        max_num_levels = [ob.num_levels for ob in self._point_sets]
+        # get the combinations of levels
+        index_composition = []
+        for sum_of_levels in range(
+            self.num_dimensions, self.num_dimensions + max_num_level
+        ):
+            index_composition.extend(
+                list(
+                    generate_compositions(
+                        sum_of_levels, self.num_dimensions, include_zero=False
+                    )
+                )
+            )
+        index_composition = numpy.array(index_composition) - 1
+        if min(max_num_levels) != max_num_level:
+            # only check if point sets have different number of levels
+            valid_comb = numpy.all(
+                numpy.greater_equal(max_num_levels, index_composition), axis=1
+            )
+            index_composition = index_composition[valid_comb, :]
+        for index_comp in index_composition:
+            level_composition_index = [
+                self._point_sets[d].level(index) for d, index in enumerate(index_comp)
+            ]
+            grid_points_ = numpy.array(
+                numpy.meshgrid(*level_composition_index)
+            ).T.reshape(-1, self.num_dimensions)
+            if grid_points is None:
+                grid_points = grid_points_
+            else:
+                grid_points = numpy.concatenate((grid_points, grid_points_), axis=0)
```

Comment:
Similarly, some commenting here would be helpful. I think this is taking the tensor products of the relevant levels?

I would in general look at whether you can rename some variables here based on our new notation to make things clearer. For example, `grid_points` probably is not meaningful and should just be `points`. Similarly, `index_composition`, `level_composition_index`, etc. might be named more clearly.

Reply:
Yeah. For each combination of levels, it grabs the list of points associated with the level number and then uses meshgrid to make the tensor products.